### PR TITLE
Bowling project

### PR DESCRIPTION
**Pull request by jakjan95**
Authors: @jakjan95 @MaciejSalamonski  @McKucia  @Pawellos

### CODE REVIEW

File: bowling/src/DirectoryHandler.cpp [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    try {
+        processData();
+    } catch (std::exception& e) {
+        std::cerr << "Cannot process directory data!\n";
+        throw;
+    }
+}
+
+void DirectoryHandler::processData() {
+    std::regex pattern(R"((\w+)/([a-zA-Z]+)(\d+).?(\w+)?)");
```
Comment (WBigaj) [inline bowling/src/DirectoryHandler.cpp:20]: I think this regex can be const

**Review by WBigaj — COMMENTED** [review]
I have checked your code and it looks solid 👍  I have only one point.

File: bowling/src/DirectoryHandler.cpp (hunk 1 above)
Comment (MaciejSalamonski) [reply bowling/src/DirectoryHandler.cpp:20]: Thanks :)

**Review by MaciejSalamonski — APPROVED** [review]

File: bowling/tests/DirectoryHandler.ut.cpp [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+    std::vector<std::string> lane1{"Name1:X|4-|3",
+                                   "Name2:34|X|0-",
+                                   ":X|22|33"};
+
+    auto isContainsLine1 = std::all_of(handlerDataFirstLane.cbegin(), handlerDataFirstLane.cend(), [lane1](const auto& el) {
+        return std::find(lane1.cbegin(), lane1.cend(), el) != lane1.cend();
+    });
+
+    ASSERT_TRUE(isContainsLine1);
```
Comment (ziobron) [inline bowling/tests/DirectoryHandler.ut.cpp:33]: there is a vector matcher ElementsAre in gtest

File: bowling/src/Player.cpp [hunk 3]
```diff
@@ -0,0 +1,36 @@
+#include "../include/Player.hpp"
+#include <iostream>
+
+int Player::getScore() {
+    int score = 0;
+    std::vector<std::pair<int, int>> twoThrows;
+
+    for (auto i = 0; i < points_.size() - 1; i += 2) {
+        twoThrows.push_back(std::make_pair(points_[i], points_[i + 1]));
```
Comment (ziobron) [inline bowling/src/Player.cpp]:
```suggestion
        twoThrows.emplace_back(points_[i], points_[i + 1]);
```

File: bowling/src/Player.cpp [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    }
+    int extraBall = twoThrows.size();
+    int throwBalls = 9 < twoThrows.size() ? 9 : twoThrows.size();
```
Comment (ziobron) [inline bowling/src/Player.cpp]:
```suggestion
    int throwBalls = (9 < twoThrows.size() ? 9 : twoThrows.size());
```
add parentheses for clarity. Or use std::max/min

File: bowling/tests/Player.ut.cpp [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    constexpr int scoreWithStrikeAfterSpare = 30;
+    ASSERT_EQ(playerTest.getScore(), scoreWithStrikeAfterSpare);
+}
+
+TEST_F(PlayerTest, ShouldCountOfScoreForGameWithAllFives) {
+    score = {5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 0};
+    Player playerTest("name", score);
+    constexpr int scoreWithAllFives = 150;
+    ASSERT_EQ(playerTest.getScore(), scoreWithAllFives);
+}
```
Comment (ziobron) [inline bowling/tests/Player.ut.cpp]: generally, you could have used parametric tests here

File: bowling/include/DataParser.hpp [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+#include <string>
+#include <vector>
+
+namespace {
+constexpr uint8_t strike = 10;
+constexpr uint8_t zeroPoints = 0;
+constexpr uint8_t addZeroPointsToVectorEndForThisSize = 25;
+constexpr char isStrike = 'X';
+constexpr char isMiss = '-';
+constexpr char isSpare = '/';
```
Comment (ziobron) [inline bowling/include/DataParser.hpp]: `is*` in name suggest me that it is a function returning a bool. Don't use verbs for variables.
`convertCharToInt` below as well.

File: bowling/include/DataParser.hpp [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+class DataParser {
+public:
+    DataParser() = default;
+    std::pair<std::string, std::vector<int>> dataParsing(std::string& processingString);
+
+private:
+    std::string stringPreparing(std::string& playerResults);
+    std::vector<std::string> stringSplitting(std::string& preparedStringWithPlayerResults);
+    std::vector<int> stringProcessing(std::vector<std::string>& vectorWithResultsForEachLine);
```
Comment (ziobron) [inline bowling/include/DataParser.hpp]: name functions with verbs, like
parseData
prepareString
splitString
processString

File: bowling/src/DataParser.cpp [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+    return vectorWithResultsForEachLine;
+}
+
+std::vector<int> DataParser::stringProcessing(std::vector<std::string>& vectorWithResultsForEachLine) {
+    std::vector<int> vectorWithResults{};
+
+    std::find_if(vectorWithResultsForEachLine.cbegin(), vectorWithResultsForEachLine.cend(),
+                 [&vectorWithResults](std::string bowlingLineString) {
+                     std::find_if(bowlingLineString.cbegin(), bowlingLineString.cend(),
```
Comment (ziobron) [inline bowling/src/DataParser.cpp:31]: using emplace_back in your lambdas tells me that std::find_if is not a proper algorithm for that.

File: bowling/tests/DataParser.ut.cpp [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+        dataParsingTuple{{"Michael:X|7/|9-|X|-8|8/|-6|X|X|X||81"},
+                         {{"Michael"}, {10, 0, 7, 3, 9, 0, 10, 0, 0, 8, 8, 2, 0, 6, 10, 0, 10, 0, 10, 0, 8, 1}}},
+        dataParsingTuple{{"ImSpecialCase:X|7/|9-|X|-8|8/|-6|X|X|X||XX5"},
+                         {{"ImSpecialCase"}, {10, 0, 7, 3, 9, 0, 10, 0, 0, 8, 8, 2, 0, 6, 10, 0, 10, 0, 10, 0, 10, 0, 10, 0, 5, 0}}},
+        dataParsingTuple{{"Radek:9-|9-|9-|9-|9-|9-|9-|9-|9-|9-||"},
+                         {{"Radek"}, {9, 0, 9, 0, 9, 0, 9, 0, 9, 0, 9, 0, 9, 0, 9, 0, 9, 0, 9, 0}}},
+        dataParsingTuple{{"Waleczny_Ziobro_69_PL:X|X|X|X|X|X|X|X|X|X||XX"},
+                         {{"Waleczny_Ziobro_69_PL"}, {10, 0, 10, 0, 10, 0, 10, 0, 10, 0, 10, 0, 10, 0, 10, 0, 10, 0, 10, 0, 10, 0, 10, 0}}},
+        dataParsingTuple{{"|----------4 cm-----------|:5/|5/|5/|5/|5/|5/|5/|5/|5/|5/||5"},
+                         {{"|----------4 cm-----------|"}, {5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5}}},
```
Comment (ziobron) [inline bowling/tests/DataParser.ut.cpp]: nice test case

File: bowling/src/BowlingLane.cpp [hunk 10]
```diff
@@ -0,0 +1,29 @@
+#include "BowlingLane.hpp"
+
+BowlingLane::GameState BowlingLane::checkGameState() {
+    for (const auto& player : players_) {
```
Comment (ziobron) [inline bowling/src/BowlingLane.cpp]: There is a loop over players but you always return after the first player check. What if the last player has not finished yet while the first has? Add a proper test case.

File: bowling/tests/BowlingLane.ut.cpp [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+TEST_F(BowlingLaneTest, ShouldWriteOutDataUsingOstreamOperator){
+    std::stringstream os;
+    os << bowlingLane_;
+    int laneNumber = bowlingLane_.getLaneNumber();
+    auto gameState = bowlingLane_.convertGameState(bowlingLane_.checkGameState());
+
+    std::string expected("### Lane " + std::to_string(laneNumber) + ": " + gameState + " ###\n");
+    ASSERT_EQ(os.str(), expected);
+}
+
```
Comment (ziobron) [inline bowling/tests/BowlingLane.ut.cpp:64]: more testcases for game state and output are needed.
What about empty files, files with many players?

File: bowling/include/Game.hpp [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+class Game {
+public:
+    Game(const std::string& directoryPath, const std::string& resultOutput = "")
+        : directoryHandler_(directoryPath), resultOutput_(resultOutput) {
+        processData();
+    }
+
+    std::vector<std::shared_ptr<BowlingLane>> getLanes() const { return lanes_; }
+    void printResult();
+    void saveDataToFile();
```
Comment (ziobron) [inline bowling/include/Game.hpp:21]: how to test whole game results?

you should have a function returning a string that is later either displayed on-screen or saved into file.
Maybe processData should return it?

File: bowling/tests/Game.ut.cpp [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+    if(testFile.is_open()){
+        while(getline(testFile,line)){
+            os << line;
+        }
+        testFile.close();
+    } else {
+        std::cout << "Cannot open file\n";
+    }
+    ASSERT_EQ(os.str(), expected);
+}
```
Comment (ziobron) [inline bowling/tests/Game.ut.cpp:95]: What about empty directories, dirs with one or many files, with empty files? They may be tested in DirectoryHandler but here some general test of some system should be provided. You can mock some functions to return proper values if you don't want to use prepared files (it is recommended way).

File: bowling/main.cpp [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+            inputDirectoryName = argv[1];
+            Game game{inputDirectoryName, resultOutputFileName};
+            game.printResult();
+            game.saveDataToFile();
+        } else {
+            std::cout << "Incorrect use, type --h or -help for help\n";
+        }
+    } catch (std::exception& e) {
+        std::cerr << "Error: " << e.what() << '\n';
+    }
```
Comment (ziobron) [inline bowling/main.cpp:42]: it's not tested.

**Review by ziobron — CHANGES_REQUESTED** [review]
bad naming conventions in data parser. Some more tests are recommended.

File: bowling/include/Game.hpp (hunk 12 above)
Comment (jakjan95) [reply bowling/include/Game.hpp:21]: For this i created function getGameResult() which is creating a string of game result output, putting that functionality into processData will be against single function responsibility(processData is responsible for filling a game data from clean string input)

File: bowling/tests/Game.ut.cpp (hunk 13 above)
Comment (jakjan95) [reply bowling/tests/Game.ut.cpp:95]: Tests added by using prepared files and directories, unfortunately github don't allow to add empty directory - so tests which dealing with empty directory I commented out to pass github CI.

File: bowling/include/DataParser.hpp (hunk 6 above)
Comment (MaciejSalamonski) [reply bowling/include/DataParser.hpp]: Done

File: bowling/include/DataParser.hpp (hunk 7 above)
Comment (MaciejSalamonski) [reply bowling/include/DataParser.hpp]: Done

File: bowling/src/DataParser.cpp (hunk 8 above)
Comment (MaciejSalamonski) [reply bowling/src/DataParser.cpp:31]: I dont quite understand why. Is the point that std::find_if is only to be used for searching? What other algorithm do you suggest? Std::find_if was the first that came to my mind :(.

File: bowling/tests/DataParser.ut.cpp (hunk 9 above)
Comment (MaciejSalamonski) [reply bowling/tests/DataParser.ut.cpp]: Thanks :)

File: bowling/src/DataParser.cpp (hunk 8 above)
Comment (ziobron) [reply bowling/src/DataParser.cpp:31]: maybe some `transform / generate / copy / move / fill`? If you want to put some data into a container it is first what comes into my mind. There is also `search` algorithm.
You have nested find_if, maybe one of them is good. Maybe even both are good, but there is something "dark" that I see find_if + emplace_back :D

File: bowling/tests/BowlingLane.ut.cpp [hunk 15]
```diff
@@ -1,6 +1,47 @@
...
+    std::vector<int> gameInProgress = {1, 2, 3};
+    players.emplace_back(std::make_shared<Player>(playerName, gameInProgress));
+    BowlingLane bowlingLane{1, players};
+    ASSERT_EQ(bowlingLane.checkGameState(), BowlingLane::GameState::IN_PROGRESS);
+}
+
+TEST(BowlingLane, BowlingLaneStateShouldReturnFinished) {
+    std::vector<std::shared_ptr<Player>> players;
+    std::vector<int> gameInProgress(20, 1);
+    players.emplace_back(std::make_shared<Player>(playerName, gameInProgress));
```
Comment (ziobron) [inline bowling/tests/BowlingLane.ut.cpp:41]: what about mixed case? first player has finished and the last one not?

File: bowling/src/Game.cpp [hunk 16]
```diff
@@ -1 +1,43 @@
...
+        } else {
+            std::cout << "Cannot open file\n";
+        }
+    } else {
+        std::cout << "Cannot save to file - no argument provided\n";
+    }
+}
+
+void Game::processData() {
+    auto parser = std::make_shared<DataParser>();
```
Comment (ziobron) [inline bowling/src/Game.cpp]: why dynamic allocation? You could create it on stack.

File: bowling/tests/Game.ut.cpp [hunk 17]
```diff
@@ -1,6 +1,124 @@
...
+    std::string line;
+    if (testFile.is_open()) {
+        while (getline(testFile, line)) {
+            os << line << '\n';
+        }
+        testFile.close();
+    } else {
+        FAIL() << "Cannot open file\n";
+    }
+    ASSERT_EQ(os.str(), expectedManyFiles);
```
Comment (ziobron) [inline bowling/tests/Game.ut.cpp]: duplication, extract it into a function

File: bowling/tests/Game.ut.cpp [hunk 18]
```diff
@@ -1,6 +1,124 @@
...
+            os << line << '\n';
+        }
+        testFile.close();
+    } else {
+        FAIL() << "Cannot open file\n";
+    }
+    ASSERT_EQ(os.str(), expectedEmptyFile);
+}
+
+//Test intentionally commented out to pass CI - The reason is that we can't crete empty directory on github repo
```
Comment (ziobron) [inline bowling/tests/Game.ut.cpp]: ok. In gtest you can add DISABLED_ before the test name. It will not execute this test, but assure that it compiles.

**Review by ziobron — CHANGES_REQUESTED** [review]

File: bowling/main.cpp (hunk 14 above)
Comment (jakjan95) [reply bowling/main.cpp:42]: Added relevant tests for results and input checking so most of this code is tested behind boundaries.

**Review by ziobron — APPROVED** [review]
Well done 🥇

**Comment by ziobron** [comment]
{
"requirement_ids": [4595, 4595, 4595, 4595, 4595, 4595, 4595, 1120],
"points": [10, 10, 10, 10, 10, 10, 10, 5],
"users": ["jakjan95", "MaciejSalamonski", "McKucia", "Pawellos"]
}
